endorsedLibraries/kotlinx.cli/src/main/kotlin/kotlinx/cli/Arguments.kt

```diff
@@ -0,0 +1,173 @@
+/*
+ * Copyright 2010-2019 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license
+ * that can be found in the LICENSE file.
+ */
+package kotlinx.cli
+
+import kotlin.reflect.KProperty
+
+internal data class CLIEntityWrapper(var entity: CLIEntity<*>? = null)
+
+/**
+ * Command line entity.
+ *
+ * @property owner parser which owns current entity.
+ */
+abstract class CLIEntity<TResult> internal constructor(internal val owner: CLIEntityWrapper) {
+    /**
+     * CLI element value.
+     */
+    internal lateinit var cliElement: ArgumentValueDelegate<TResult>
+    /**
+     * Wrapper  for element - read only property.
+     * Needed to close set of variable [cliElement].
+     */
+    val delegate: ArgumentValueDelegate<TResult>
+        get() = cliElement
+
+    /**
+     * Value of entity.
+     */
+    var value: TResult
+        get() = cliElement.value
+        set(value) = delegate.setValue(null, ::value, value)
```

Comment:
`delegate.setValue` doesn't use `thisRef` and `property` parameters, so maybe we could provide the way to set value without them, for example by making its `value` a var property. This way we could avoid creating `::value` property reference.

Reply:
If I understood your suggestion right way, I have a concern that ValueOrigin isn't set right or I should set it also here. It seems to have all in one place, isn't it?

Reply:
Fixed